onnxruntime/contrib_ops/rocm/bert/fast_gelu_impl.cu

```diff
@@ -160,41 +131,49 @@ template <>
-      const int n = input_length / 4;
-      const int gridSize = (n + blockSize - 1) / blockSize;
-      const float2* input4 = reinterpret_cast<const float2*>(input);
-      const float2* bias4 = reinterpret_cast<const float2*>(bias);
-      float2* output4 = reinterpret_cast<float2*>(output);
-      if (bias == nullptr)
-        hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel4<blockSize>), dim3(gridSize), dim3(blockSize), 0,
-                           stream, A2, B2, C2, one2, two2, n, input4, output4);
-      else
-        hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel4Bias<blockSize>), dim3(gridSize), dim3(blockSize), 0,
-                           stream, A2, B2, C2, one2, two2, n, bias_length / 4, input4, bias4, output4);
-    } else {
-      const int n = input_length / 2;
-      const int gridSize = (n + blockSize - 1) / blockSize;
-      const half2* input2 = reinterpret_cast<const half2*>(input);
-      const half2* bias2 = reinterpret_cast<const half2*>(bias);
-      half2* output2 = reinterpret_cast<half2*>(output);
-      hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel2<blockSize>), dim3(gridSize), dim3(blockSize), 0,
-                         stream, A2, B2, C2, one2, two2, n, bias_length / 2, input2, bias2, output2);
-    }
+  const int gridSize = (input_length + blockSize - 1) / blockSize;
+  const half oneT = half(one);
+  const half twoT = half(two);
+  if (use_half2 && prop.major >= 7) {
+      if (bias != nullptr) {
+        if (0 == (bias_length % 8) && (input_length >= 3145728)) {
+          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluBiasKernelVec<half, blockSize, 8>), dim3(gridSize),
+                                             dim3(blockSize), 0, stream, A, B, C, oneT, twoT, input_length, bias_length,
+                                             input, bias, output);
+        } else if (0 == (bias_length % 4)) {
+          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluBiasKernelVec<half, blockSize, 4>), dim3(gridSize),
+                                             dim3(blockSize), 0, stream, A, B, C, oneT, twoT, input_length, bias_length,
+                                             input, bias, output);
+        } else if (0 == (bias_length % 2)) {
+          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluBiasKernelVec<half, blockSize, 2>), dim3(gridSize),
+                                             dim3(blockSize), 0, stream, A, B, C, oneT, twoT, input_length, bias_length,
+                                             input, bias, output);
+        } else {
+          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel<half, blockSize>), dim3(gridSize), dim3(blockSize), 0,
+                                             stream, A, B, C, oneT, twoT, input_length, bias_length, input, bias, output);
+        }
+      } else {
+        if (0 == (input_length % 8) && (input_length >= 3145728)) {
+          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernelVec<half, blockSize, 8>), dim3(gridSize),
```

Comment:
Should grid_size be reduced to 1/8 if we use a vector size of 8?

Reply:
Yes, the grid_size should be reduced to 1/ILP. Thanks.